File: sdk/python/examples/rdt3_receiver.py
from tcp_lab import TransportProtocol, SystemContext, Packet, TcpHeader, TcpFlags

class Rdt3Receiver(TransportProtocol):
    def __init__(self):
        self.expected_seq = 0

    def init(self, ctx: SystemContext) -> None:
        self.expected_seq = 0

    def on_app_data(self, ctx: SystemContext, data: bytes) -> None:
        pass

    def on_packet(self, ctx: SystemContext, packet: Packet) -> None:
        h = packet.header
        
        if h.seq_num == self.expected_seq:
            ctx.log(f"Received correct packet {self.expected_seq}")
            ctx.deliver_data(packet.payload)
            self.send_ack(ctx, self.expected_seq)
            self.expected_seq = 1 - self.expected_seq
        else:
            ctx.log(f"Duplicate/Out-of-order packet {h.seq_num}, expected {self.expected_seq}")
            self.send_ack(ctx, 1 - self.expected_seq)

    def on_timer(self, ctx: SystemContext, timer_id: int) -> None:
        pass

    def send_ack(self, ctx: SystemContext, ack_num: int) -> None:
        h = TcpHeader(ack_num=ack_num, flags=TcpFlags.ACK)
        ctx.send_packet(Packet(h, b""))

File: sdk/rust/src/examples/rdt3_receiver.rs
use tcp_lab_core::{Packet, SystemContext, TransportProtocol};

pub struct Rdt3Receiver {
    expected_seq: u32,
}

impl Rdt3Receiver {
    pub fn new() -> Self {
        Self { expected_seq: 0 }
    }
}

impl TransportProtocol for Rdt3Receiver {
    fn init(&mut self, ctx: &mut dyn SystemContext) {
        ctx.log("Rdt3Receiver initialized");
    }

    fn on_packet(&mut self, ctx: &mut dyn SystemContext, packet: Packet) {
        if packet.header.seq_num == self.expected_seq {
            ctx.log(&format!("Received expected packet seq {}", packet.header.seq_num));
            ctx.deliver_data(&packet.payload);
            
            let ack = Packet::new_ack(0, self.expected_seq, 100);
            ctx.send_packet(ack);
            
            self.expected_seq += 1;
        } else {
            ctx.log(&format!("Received unexpected packet seq {}, expected {}", packet.header.seq_num, self.expected_seq));
            // Resend ACK for last correctly received packet (if any)
            if self.expected_seq > 0 {
                let ack = Packet::new_ack(0, self.expected_seq - 1, 100);
                ctx.send_packet(ack);
            }
        }
    }

    fn on_timer(&mut self, _ctx: &mut dyn SystemContext, _timer_id: u32) {}

    fn on_app_data(&mut self, _ctx: &mut dyn SystemContext, _data: &[u8]) {}
}